In [24]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import re


In [25]:
df = pd.read_csv('googleplaystore.csv')
df.head()



App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

## Knowing Dataset

In [26]:
df.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [27]:
df.shape

(10841, 13)

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [29]:
df.dtypes

App                object
Category           object
Rating            float64
Reviews            object
Size               object
Installs           object
Type               object
Price              object
Content Rating     object
Genres             object
Last Updated       object
Current Ver        object
Android Ver        object
dtype: object

In [30]:
df.isnull().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64

## Cleaning Dataset

#### Task 1: Lower case columns name

In [31]:
df.columns = [x.lower() for x in df.columns]
df.columns

Index(['app', 'category', 'rating', 'reviews', 'size', 'installs', 'type',
       'price', 'content rating', 'genres', 'last updated', 'current ver',
       'android ver'],
      dtype='object')

#### Task 2: Change column date type to numrical.

Below columns contains the numerical data type but in the form of object. In this task will be change it into interger/float dtypes.

But we change it's datatyep, we will remove and replace data where it is necessary.

In [32]:
df[['reviews','size','installs','price','current ver', 'android ver']]

reviews                size     installs price         current ver  \
0         159                 19M      10,000+     0               1.0.0   
1         967                 14M     500,000+     0               2.0.0   
2       87510                8.7M   5,000,000+     0               1.2.4   
3      215644                 25M  50,000,000+     0  Varies with device   
4         967                2.8M     100,000+     0                 1.1   
...       ...                 ...          ...   ...                 ...   
10836      38                 53M       5,000+     0                1.48   
10837       4                3.6M         100+     0                 1.0   
10838       3                9.5M       1,000+     0                 1.0   
10839     114  Varies with device       1,000+     0  Varies with device   
10840  398307                 19M  10,000,000+     0  Varies with device   

              android ver  
0            4.0.3 and up  
1            4.0.3 and up  
2            4.0.3 and up  
3              4.2 and up  
4              4.4 and up  
...                   ...  
10836          4.1 and up  
10837          4.1 and up  
10838          2.2 and up  
10839  Varies with device  
10840  Varies with device  

[10841 rows x 6 columns]

#### Reviews columns

In [33]:
df[df['reviews'].str.contains('[A-Z]', flags=re.I)]

app category  rating reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         size installs type     price content rating             genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      last updated current ver android ver  
10472       1.0.19  4.0 and up         NaN

As you can see, at index 10472, there is 3.0M reviews. Which can't be true. With careful observation we can see that, the apps category type is missing in this row and other columns value is push to the next left column.

If we check this particular apps in Google play station.It's fall under 'Lifestyle' category.

Now, We will update with category and other respected columns. 

In [34]:
result  = df.iloc[10472].values.tolist()
result.insert(1, 'Lifestyle')
result.pop()
df.iloc[10472] = result
df.iloc[10472]

app               Life Made WI-Fi Touchscreen Photo Frame
category                                        Lifestyle
rating                                                1.9
reviews                                                19
size                                                 3.0M
installs                                           1,000+
type                                                 Free
price                                                   0
content rating                                   Everyone
genres                                                NaN
last updated                            February 11, 2018
current ver                                        1.0.19
android ver                                    4.0 and up
Name: 10472, dtype: object

In [35]:
df['reviews'] = df['reviews'].astype(int)

#### Size column

In 'size' columns. There is no specific size for few apps. Of this analysis purpose we will consider those apps having 10.0M size.

In [36]:
df['size'] = df['size'].apply(lambda x : '10.0' if x == 'Varies with device' else x.replace('M', ""))
df['size'] = df['size'].apply(lambda x : x.replace('K', ""))
df['size'] = df['size'].apply(lambda x : x.replace('k', ""))
df['size'].astype(float)

0        19.0
1        14.0
2         8.7
3        25.0
4         2.8
         ... 
10836    53.0
10837     3.6
10838     9.5
10839    10.0
10840    19.0
Name: size, Length: 10841, dtype: float64

#### Installs Column.

In [37]:
df['installs'].unique()

array(['10,000+', '500,000+', '5,000,000+', '50,000,000+', '100,000+',
       '50,000+', '1,000,000+', '10,000,000+', '5,000+', '100,000,000+',
       '1,000,000,000+', '1,000+', '500,000,000+', '50+', '100+', '500+',
       '10+', '1+', '5+', '0+', '0'], dtype=object)

In [38]:
df['installs'] = df['installs'].apply(lambda x : x.replace('+', ''))
df['installs'] = df['installs'].apply(lambda x : x.replace(',', ''))
df['installs'].astype(int)

0           10000
1          500000
2         5000000
3        50000000
4          100000
           ...   
10836        5000
10837         100
10838        1000
10839        1000
10840    10000000
Name: installs, Length: 10841, dtype: int64

#### Price column.

In [39]:
df['price'] = df['price'].apply(lambda x : x.replace('$', ''))
df['price'].astype(float)

0        0.0
1        0.0
2        0.0
3        0.0
4        0.0
        ... 
10836    0.0
10837    0.0
10838    0.0
10839    0.0
10840    0.0
Name: price, Length: 10841, dtype: float64

In [40]:
df['price'].unique()

array(['0', '4.99', '3.99', '6.99', '1.49', '2.99', '7.99', '5.99',
       '3.49', '1.99', '9.99', '7.49', '0.99', '9.00', '5.49', '10.00',
       '24.99', '11.99', '79.99', '16.99', '14.99', '1.00', '29.99',
       '12.99', '2.49', '10.99', '1.50', '19.99', '15.99', '33.99',
       '74.99', '39.99', '3.95', '4.49', '1.70', '8.99', '2.00', '3.88',
       '25.99', '399.99', '17.99', '400.00', '3.02', '1.76', '4.84',
       '4.77', '1.61', '2.50', '1.59', '6.49', '1.29', '5.00', '13.99',
       '299.99', '379.99', '37.99', '18.99', '389.99', '19.90', '8.49',
       '1.75', '14.00', '4.85', '46.99', '109.99', '154.99', '3.08',
       '2.59', '4.80', '1.96', '19.40', '3.90', '4.59', '15.46', '3.04',
       '4.29', '2.60', '3.28', '4.60', '28.99', '2.95', '2.90', '1.97',
       '200.00', '89.99', '2.56', '30.99', '3.61', '394.99', '1.26',
       '1.20', '1.04'], dtype=object)

### current ver column

In [41]:
df[df['current ver'].isna()==True]

app             category rating  reviews  \
15     Learn To Draw Kawaii Characters       ART_AND_DESIGN    3.2       55   
1553              Market Update Helper   LIBRARIES_AND_DEMO    4.1    20145   
6322            Virtual DJ Sound Mixer                TOOLS    4.2     4010   
6803                         BT Master               FAMILY    NaN        0   
7333                       Dots puzzle               FAMILY      4      179   
7407                   Calculate My IQ               FAMILY    NaN       44   
7730                            UFO-CQ                TOOLS    NaN        1   
10342                   La Fe de Jesus  BOOKS_AND_REFERENCE    NaN        8   

      size installs  type price content rating             genres  \
15     2.7     5000  Free     0       Everyone       Art & Design   
1553    11  1000000  Free     0       Everyone   Libraries & Demo   
6322   8.7   500000  Free     0       Everyone              Tools   
6803   222      100  Free     0       Everyone          Education   
7333    14    50000  Paid  0.99       Everyone             Puzzle   
7407   7.2    10000  Free     0       Everyone      Entertainment   
7730   237       10  Paid  0.99       Everyone              Tools   
10342  658     1000  Free     0       Everyone  Books & Reference   

            last updated current ver android ver  
15          June 6, 2018         NaN  4.2 and up  
1553   February 12, 2013         NaN  1.5 and up  
6322        May 10, 2017         NaN  4.0 and up  
6803    November 6, 2016         NaN  1.6 and up  
7333      April 18, 2018         NaN  4.0 and up  
7407       April 3, 2017         NaN  2.3 and up  
7730        July 4, 2016         NaN  2.0 and up  
10342   January 31, 2017         NaN  3.0 and up

There are 8 NaN in current ver column. We will assign 1.0 to each of them.

In [42]:
df['current ver'] = df['current ver'].fillna(1.0)

In [43]:
df[df['current ver'] == 'Varies with device'].shape

(1459, 13)

There are 1459 rows where current version of apps 'Varies with device'. We will assign 1.0 for them as well.

In [44]:
df['current ver'] = df['current ver'].apply(lambda x : 1.0 if x == 'Varies with device' else x)

In [45]:
def to_float(x):
    if '.' in x:
        result = x.split('.')
        return result[0]+ "." + result[1]
    else:
        return x + '.' + '0'
        
    
df['current ver'] = df['current ver'].apply(lambda x : to_float(str(x)))


In [47]:
df['current ver'].unique()

array(['1.0', '2.0', '1.2', '1.1', '6.1', '2.9', '2.8', '3.8', '3.1',
       '2.2', '5.5', '4.0', '1.5', '1.03', '6.0', '6.7', '2.20', '1.6',
       '2.1', '1.3', '1.46', '11.0', '3.0', '1.7', '2.5', '2.493', '1.9',
       '2.20 Build 02', '1.37', '0.2', '4.47', '1.79', '2.3', '8.31',
       '10.0', '1.10', '3.20', '1.4', '4.6', '1.40', '2.17', '6.10',
       '5.4', '8.0', '3.3', '1.20', '18.5', '1.15', '3.2', '1.911805270',
       '9.1', '15.0', '18.05', '5.0', '3.12', '1.28', '9.00', '14.0',
       '3.05', '7.0', '3.9', '3.4', '2.19', '9.7', '4.9', '5.2', '3.550',
       '7.3', '5.1', '4.1', '99.0', '12.2', '1.12', '0.9', '3.24', '18.7',
       '3.19', '6.4', '3.7', '2.6', '5.20', '4.3', '10.46', '11.1',
       '2.22', '2.11', '1.08', '6.3', '4.7', '3.16', '4.2', '1.01', '2.4',
       '1.8', '37.0', '5.29', '9.8', '11.8', '24.6', '47.1', 'R6.0',
       '69.0', '1.60', '11.4', '2.18', '7.5', '37.1', '12.8', '9.9',
       '42.0', '0.1', '7.3(800241)', '7.73', '4.98', '7.21', '3.47',
  